ui/apps/platform/src/services/RolesService.ts

```diff
@@ -79,13 +80,32 @@ export function fetchUserRolePermissions(): Promise<{ response: Role }> {
 
 const permissionSetsUrl = '/v1/permissionsets';
 
+export type Traits = {
+    mutabilityMode: MutabilityMode;
+    origin?: Origin;
+};
+
+export type MutabilityMode = 'ALLOW_MUTATE' | 'ALLOW_MUTATE_FORCED';
+export type Origin = 'IMPERATIVE' | 'DECLARATIVE' | 'DEFAULT';
```

Comment:
Because TypeScript does not scope these types within `Traits` like the corresponding proto file, I often prepend it like a prefix to reduce possible ambiguity in code that import the types:
* `TraitsMutabilityMode`
* `TraitsOrigin`

Reply:
Recommended as a followup after you merge, because AuthService and RolesService both depend on these types, please factor out so frontend file structure is more parallel to backend file structure:
* backend: https://github.com/stackrox/stackrox/blob/master/proto/storage/traits.proto
* frontend: src/types/traits.proto.ts if type is in a storage proto file, but in frontend services file if (as is more and more often) backend type is in backend service proto file

Reply:
Having recommended to factor out `Traits` on the one side of the coin, here is a question on the other side of the coin. Do the optionality of properties of `traits` property in responses depend on the requests?

For example, I noticed below `traits == null` which seems slightly inconsistent with `traits?: Traits` because that would mean `traits?: Traits | null`

So, if there is a non-nullish `traits` property in response for access control requests, is `origin` property required?

Reply:
> Do the optionality of properties of traits property in responses depend on the requests?
> So, if there is a non-nullish traits property in response for access control requests, is origin property required?

Currently, `traits` are treated as an optional API request field for auth providers, groups, roles etc. That means that is possible to create role with empty `traits`, or role with `traits` and no `origin`. For simplicity, the `traits` are not filled with default value on the backend, meaning we should correctly handle resources with empty `traits` or empty `origin`.

> For example, I noticed below traits == null which seems slightly inconsistent with traits?: Traits because that would mean traits?: Traits | null

I don't entirely get what is the inconsistency - perhaps because I'm still struggling to learn how all the different ways of saying "there is no value" in Typescript work :)

Reply:
At first I misread `traits == null` as `traits === null` but then understood the `==` idiom for either
* `undefined` which corresponds to `?` in `traits?: Traits`
* `null` which I did see in the response for some request elsewhere in access control, therefore `traits?: Traits | null` describes that either absence of property or presence of property with `null` value are possible.

Benefit of being as precise as possible on this detail:
* The backend proto in this case is not clear that `null` or `undefined` are possible.
* It reduces the risk that engineers (either backend or frontend) who are not part of the original feature team, who might need to modify the code later, might write less robust predicates than your helper functions. By the way, factoring out the helper functions like you did is a huge help to make the code maintainable.